neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/dialect/impl/RemoteNeo4jSequenceGenerator.java

```diff
@@ -0,0 +1,253 @@
+	private void getUpdateTableSequenceQuery(Statements statements, IdSourceKeyMetadata idSourceKeyMetadata, String sequenceName, int initialValue, int increment) {
+		String acquireLockQuery = acquireLockQuery( idSourceKeyMetadata, initialValue, increment );
+		Statement acquireLockStatement = new Statement( acquireLockQuery, params( sequenceName, initialValue, increment ) );
+		acquireLockStatement.setResultDataContents( Arrays.asList( Statement.AS_ROW ) );
+		statements.addStatement( acquireLockStatement );
+
+		String updateQuery = increaseQuery( idSourceKeyMetadata, increment );
+		Statement updateStatement = new Statement( updateQuery, params( sequenceName, initialValue, increment ) );
+		updateStatement.setResultDataContents( Arrays.asList( Statement.AS_ROW ) );
+		statements.addStatement( updateStatement );
+	}
+
+	protected String increaseQuery(IdSourceKeyMetadata idSourceKeyMetadata, int increment) {
+		StringBuilder queryBuilder = new StringBuilder();
+		Label generatorKeyLabel = DynamicLabel.label( idSourceKeyMetadata.getName() );
+		queryBuilder.append( " MATCH (n " );
+		queryBuilder.append( labels( generatorKeyLabel.name(), NodeLabel.TABLE_BASED_SEQUENCE.name() ) );
+		queryBuilder.append( " { " );
+		queryBuilder.append( idSourceKeyMetadata.getKeyColumnName() );
+		queryBuilder.append( ": {" );
+		queryBuilder.append( SEQUENCE_NAME_QUERY_PARAM );
+		queryBuilder.append( "}} )" );
+		queryBuilder.append( " SET n." );
+		queryBuilder.append( idSourceKeyMetadata.getValueColumnName() );
+		queryBuilder.append( " = n." );
+		queryBuilder.append( idSourceKeyMetadata.getValueColumnName() );
+		queryBuilder.append( " + " );
+		queryBuilder.append( increment );
+		queryBuilder.append( " REMOVE n.__locked RETURN n." );
+		queryBuilder.append( idSourceKeyMetadata.getValueColumnName() );
+		String query = queryBuilder.toString();
+		return query;
+	}
+
+	/**
+	 * Generate the next value in a sequence for a given {@link IdSourceKey}.
+	 *
+	 * @param idSourceKey identifies the generator
+	 * @param increment the difference between to consecutive values in the sequence
+	 * @param initialValue the initial value of the given generator
+	 * @return the next value in a sequence
+	 */
+	public Number nextValue(IdSourceKey idSourceKey, int increment, int initialValue) {
+		String sequenceName = sequenceName( idSourceKey );
+		Statements statements = updateNextValueQuery( idSourceKey, initialValue, increment );
+		StatementsResponse statementsResponse = neo4jDb.executeQueriesInNewTransaction( statements );
+		List<StatementResult> results = statementsResponse.getResults();
+		Integer nextValue = (Integer) results.get( 1 ).getData().get( 0 ).getRow().get( 0 );
+		// sequence nodes are expected to have been created up-front
+		if ( idSourceKey.getMetadata().getType() == IdSourceType.SEQUENCE ) {
+			if ( nextValue == null ) {
+				throw new HibernateException( "Sequence missing: " + sequenceName );
+			}
+		}
+		return nextValue.longValue() - increment;
```

Comment:
The only way I found to make it work in a multi-threaded environment is to first increment the value and then read it . Our API allows for an initial value and to make sure that I'm actually reading the correct one the first time I need to decrement the value I obtain from the db.